dvc/repo/tree.py

```diff
@@ -111,6 +146,15 @@ def walk(self, top, topdown=True):
         for out in outs:
             trie[out.path_info.parts] = out
 
+            if out.is_dir_checksum and (self.fetch or self.stream):
+                # will pull dir cache if needed
+                with self.repo.state:
+                    cache = out.collect_used_dir_cache()
```

Comment:
Might be slow to collect. You only need `out.dir_cache`, so maybe it is worth creating extracting some method to fetch .dir for it into an `out` method. But I'm ok with keeping it as is for now.